R/bploop.R

```diff
@@ -252,14 +224,21 @@ bploop.lapply <-
 ## - length of 'X' is unknown (defined by ITER())
 ## - results not pre-allocated; list grows each iteration if no REDUCE
 ## - args wrapped in arglist with different chunks from ITER()
-##
-
+## Arguments
+## - ITER: Return a list where each list element will be passed to FUN
+##   1. if nothing to proceed, it should return list(NULL)
+##   2. if the task is iterate the seed stream only, it should return
+##      an object from .rng_bploop_iter()
+## - FUN: A function that will be evaluated in the worker
+## - ARGS: the arguments to FUN
 bploop.iterate <-
-    function(manager, ITER, FUN, ARGFUN, BPPARAM, REDUCE, init, reduce.in.order,
-             ...)
+    function(manager, ITER, FUN, ARGS, BPPARAM, REDUCE, init, reduce.in.order)
 {
     cl <- bpbackend(BPPARAM)
 
+    seed <- .bpnextRNGstream(BPPARAM)
+    # seed <- .rng_next_substream(seed)
```

Comment:
The intention was that `.bpnextRNGstream()` was used on `bpstart()`, and the substreams were used for each element of `X` as mentioned in this commit bf9de5e1f01845c97d08c5b1fde8e31ce9b41eb3.

Reply:
I think the use of `.bpnextRNGstream()` in the original code was an error, it should have been just `.RNGseed()`.

Reply:
Thanks, it looks like you did it on purpose, we need to memorize the seed used by the previous apply function. Otherwise two same apply calls will give the same result. There is a unit test for it

https://github.com/Bioconductor/BiocParallel/blob/70b4d051a3f99db9c52a4bf70e88ed7cc631df5a/inst/unitTests/test_rng.R#L196-L199

Since `.bpnextRNGstream` will iterate the seed and store it in the BPPARAM, it is able to pass the test, but `.RNGseed` cannot change the seed in BPPARAM and therefore cannot pass the test. 

But I do see a bug in my code caused by the fallback SerialParam in `bpinit`, I'll push a commit to fix it.

Reply:
`.bpnextRNGstream` has been removed from the commit 50651fd as a side effect of the bug fixing anyway. `bpinit` is now able to memorize the seed stream used by the fallback SerialParam and set it back to the original *Param after `bploop`, this promises the result is consistent for the *Param with different worker numbers. For example
```
## This will fallback to the SerialParam, but it is still able to iterate its seed stream
> p <- SnowParam(workers = 1, RNGseed = 1)
> bpstart(p)
> as.numeric(bplapply(1, test, BPPARAM = p))
[1] 0.7720786
> as.numeric(bplapply(1, test, BPPARAM = p))
[1] 0.4301437
``` 
I do think there is still more work for the RNG seed stream, but at least with this new version we do not have to duplicate our implementation in many functions.